sprint 7

Just-Roman opened the pull request:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager) 1.2 add method save()
2. add Class CSVFormatter: 2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException 3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task : add method Integer getEpicId()
5. interface TaskManager : add method setNextId(Integer nextId)

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)
6. Delete space

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)
6. Delete space

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,168 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+
+    private static final String HOME = System.getProperty("user.dir");
```
Пути или сам файл лучше передавать в метод `loadFromFile`, харкодить тут пути и на звание файла не нужно)

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,168 @@
+    private static final String fileName = "tasksFile.csv";
+    private static final Path TASKS_FILE_PATH = Paths.get(HOME, fileName);
+
+    public static void loadFromFile(TaskManager taskManager) {
```
В этот метод лучше передавать файл передавать файл

slavrikk replied:
Так же объект `FileBackedTaskManager` лучше создать внутри метода

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,168 @@
+
+    private void save() {
+
+        try {
```
Лучше весь код вызывать внутри одного try catch

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    @Override
+    public void updateStatusEpic(Epic epic) {
```
Для статуса это лишнее, этот метод лучше сделать приватным внутри класса `InMemoryTaskManager`

slavrikk commented on src/ru/practicum/task_tracker/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,168 @@
+                int commaIndex = row.indexOf(",");
+
+                Task task = switch (row.substring(0, commaIndex)) {
+                    case "TASK" -> taskManager.createTask(CSVFormatter.taskFromString(row));
```
Тут можно напрямую добавлять в коллекции таски, без вызова методов create

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -46,8 +45,12 @@ public Task getByTaskId(Integer taskId) {
     public Task createTask(Task task) {
-            task.setId(getNextId());
-        tasks.put(task.getId(), task);
+        Integer id = task.getId();
```
id добавляется только внутри менеджера, эта проверка лишняя

slavrikk replied:
Замечание для всех методов `create`*

slavrikk commented on src/ru/practicum/task_tracker/manager/InMemoryTaskManager.java:
```diff
@@ -215,7 +228,11 @@ public void deleteAllTasks() {
+        return nextId++;
+    }
+
+    public void setNextId(Integer nextId) {
```
Лучше изменить модификатор доступа переменной `nextId` на `protected` и ее переопределить в `FileBackedTaskManager `

slavrikk commented on src/ru/practicum/task_tracker/manager/TaskManager.java:
```diff
@@ -43,4 +43,6 @@ public interface TaskManager {
 
     void deleteAllTasks();
+
+    public void setNextId(Integer nextId);
```
Этот метод лишний

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)
6. Delete space
7. Corrected comments

Commit:
1. add Class FileBackedTaskManager extends InMemoryTaskManager :
1.1 add method loadFromFile(TaskManager taskManager)
1.2 add method save()
2. add Class CSVFormatter:
2.1 add methods:
 - toString(Task task)
- taskFromString(String csvRow)
- epicFromString(String csvRow)
- subtaskFromString(String csvRow)
- getStatus (String str)
- getHeader()
3. add Class  ManagerSaveException extends RuntimeException
3.1 add methods:
- saveException(Exception e)
- loadException(Exception e)
4. Class Task :
add method Integer getEpicId()
5. interface TaskManager :
add method setNextId(Integer nextId)
6. Delete space
7. Corrected comments
